File: 3332-minimum-operations-to-exceed-threshold-value-ii/minimum-operations-to-exceed-threshold-value-ii.cpp
class Solution {
public:
    int minOperations(vector<int>& nums, int k) {
        priority_queue<long long, vector<long long>, greater<long long>> pq;
        for(auto &it:nums){
            pq.push(it);
        }

        int ans = 0;

        while(pq.top() < k){
            long long x = pq.top();
            pq.pop();
            long long y = pq.top();
            pq.pop();
            
            long long z = min(x, y) * 2 + max(x, y);

            pq.push(z);

            ans += 1;
        }

        return ans;
    }
};

File: 3332-minimum-operations-to-exceed-threshold-value-ii/minimum-operations-to-exceed-threshold-value-ii.go
func minOperations(nums []int, k int) int {
    min_pq := &MinHeap{}
    heap.Init(min_pq)

    for _, num := range nums {
        heap.Push(min_pq, num)
    }
    ops := 0
    for min_pq.Len() >= 2 {
        x := heap.Pop(min_pq).(int)
        y := heap.Pop(min_pq).(int)

        if x >= k && y >= k {
            break
        }
        
        val := min(x, y) * 2 + max(x, y)
        ops++
        
        heap.Push(min_pq, val)
    }

    return ops
}

type MinHeap []int

func (pq MinHeap) Len() int {
    return len(pq)
}
func (pq MinHeap) Less(i, j int) bool {
    return pq[i] < pq[j]
}
func (pq MinHeap) Swap(i, j int) {
    pq[i], pq[j] = pq[j], pq[i]
}
func (pq *MinHeap) Push(x any) {
    *pq = append(*pq, x.(int))
}
func (pq *MinHeap) Pop() any {
    temp := *pq
    x := temp[len(temp)-1]
    *pq = temp[:len(temp)-1]

    return x
}